hoodie-client/src/main/java/com/uber/hoodie/func/LazyInsertIterable.java

```diff
@@ -74,30 +77,31 @@ protected void start() {
     // Executor service used for launching writer thread.
     final ExecutorService writerService = Executors.newFixedThreadPool(1);
     try {
-      // Used for buffering records which is controlled by
-      // HoodieWriteConfig#WRITE_BUFFER_LIMIT_BYTES.
-      final BufferedIterator<T, HoodieRecord<T>> bufferedIterator = new BufferedIterator<>(inputItr,
-          hoodieConfig.getWriteBufferLimitBytes(),
-          HoodieIOHandle.createHoodieWriteSchema(hoodieConfig));
-      Future<List<WriteStatus>> writerResult = writerService.submit(() -> {
-        logger.info("starting hoodie writer thread");
-        // Passing parent thread's TaskContext to newly launched thread for it to access original
-        // TaskContext
-        // properties.
-        TaskContext$.MODULE$.setTaskContext(sparkThreadTaskContext);
-        List<WriteStatus> statuses = new LinkedList<>();
-        try {
-          statuses.addAll(handleWrite(bufferedIterator));
-          logger.info("hoodie write is done; notifying reader thread");
-          return statuses;
-        } catch (Exception e) {
-          logger.error("error writing hoodie records", e);
-          bufferedIterator.markAsFailed(e);
-          throw e;
-        }
-      });
-      // Buffering records into internal buffer. This can throw exception either if reading
-      // records from spark fails or
+      // Used for buffering records which is controlled by HoodieWriteConfig#WRITE_BUFFER_LIMIT_BYTES.
+      final BufferedIterator<HoodieRecord<T>, Optional<IndexedRecord>> bufferedIterator =
+          new BufferedIterator<HoodieRecord<T>, Optional<IndexedRecord>>(inputItr,
+              hoodieConfig.getWriteBufferLimitBytes(),
+              HoodieIOHandle.createHoodieWriteSchema(hoodieConfig),
+              new HoodiePayloadFunction(HoodieIOHandle.createHoodieWriteSchema(hoodieConfig)));
+      Future<List<WriteStatus>> writerResult =
+          writerService.submit(
+              () -> {
+                logger.info("starting hoodie writer thread");
+                // Passing parent thread's TaskContext to newly launched thread for it to access original TaskContext
+                // properties.
+                TaskContext$.MODULE$.setTaskContext(sparkThreadTaskContext);
```

Comment:
`TaskContext$.MODULE$.setTaskContext(sparkThreadTaskContext);` this is the only boiler plate code here. The other code around is a little different in each scenario. Trying to avoid introducing new first class types if possible.